---
title: "Chloroplast_analysis"
author: "Julia Harenčár"
format: 
  html:
    code-fold: true
    code-tools: true
    code-overflow: wrap
    toc: true
    number-sections: true
theme:
  light: minty
  dark: superhero
editor_options: 
  chunk_output_type: console
---

## Chloroplast analysis

### Unrooted tree

```{r}
# library(rdiversity)
library(vcfR)
library(ape)
library(tidyverse)
library(ggplot2)
library(adegenet)
library(poppr)

# Read in VCF
vcf <- read.vcfR("/Users/juliaharencar/Documents/Github/Chloroplast_analysis/AVH_filtered_chloroplast.vcf.gz", convertNA  = TRUE)
unfilt_vcf <- read.vcfR("/Users/juliaharencar/Documents/Github/Chloroplast_analysis/unfiltered_chloroplast.vcf.gz", convertNA  = TRUE)

# TREE below
# working tree from my old cati code below, need to figure out if this is right/reasonable - write up methods!!
# convert to geneind
gind <- vcfR2genind(unfilt_vcf)

# calculate and plot
phylo <- aboot(gind, strata = NULL, tree = "nj", distance = "nei.dist", root = FALSE)
plot.phylo(phylo, type = 'unrooted')
plot.phylo(phylo)

### below are other attempts at making a tree
#### other maybe works below if change how plotting?
# Extract the sequence data from the VCF object:
seq_data <- extract.gt(vcf)

# Convert the sequence data to a format suitable for dist.dna:
seq_matrix <- as.matrix(seq_data)

# Calculate pairwise genetic distances with the dist.dna function from the ape package:
genetic_distances <- dist.dna(seq_matrix, model = "N")

# Calculate neighbor joining tree
nj_tree <- nj(genetic_distances)

plot(nj_tree, "unrooted", use.edge.length=TRUE, lab4ut="axial", root.edge=FALSE)

## OTHER below
# Create hierarchical clustering tree... 
hc <- hclust(genetic_distances)

# Convert hierarchical clustering tree to a phylogenetic tree
dist_tree <- as.phylo(hc)

# Plot distance tree
ggtree(dist_tree, layout = "rectangular") +
  geom_tiplab() +
  geom_treescale(width = 0.10, x = 0) +
  xlim(0, .15) +
  labs(title = "IQtree distance tree")

# Convert VCF to pairwise genetic distance matrix
dist <- gen2dist(vcf, biallelic = TRUE)


```

### Rooted tree with other local Costus

```{r}
# New VCF includes BRAC that had coverage greater or equal to 0.5x; includes all GUAN and LAEV and the few LASI - all additions from OG_plate1 sequencing

# no VCF filtering - need to play and see which filter fucked things up above.

# read in unfiltered vcf
unfilt_vcf.rooted <- read.vcfR("/Users/juliaharencar/Documents/Github/Chloroplast_analysis/unfiltered_chloroplast_w.outgroups.vcf.gz", convertNA  = TRUE)

# convert to geneind
gind <- vcfR2genind(unfilt_vcf.rooted)

# calculate and plot
phylo <- aboot(gind, strata = NULL, tree = "nj", distance = "nei.dist", root = FALSE)
plot.phylo(phylo)

```

#### Notes on Chloroplast tree results

It looks like all hybrids have a villosissimus chloroplast, which is shocking! Many hybs looked like alle, thought bc of maternal effects but must be dominance of alle floral traits especially (?) - double check stuff we called alle that turned out to be hyb - fall with vill in chloroplast tree? Where do backcrosses fall? -

### PCA
```{bash}
### My modified version for of Brock's pca code for chloroplast analysis
## SETUP
mkdir -p PCA
PREFIX="chloro_pca"
#OUT="chloro_w.outgroups"

## convert filtered vcf (see VCF_QC.qmd) to plink format
plink2 --vcf chloroplast_AHV_filtered.vcf.gz --recode transpose --out $PREFIX --allow-extra-chr

## Run PCA with bim/bam input
plink2 --tfile ${PREFIX} --pca --out ${PREFIX} --allow-extra-chr
#NOTES: not working on larger vcf with outgroups... added a missingness filter that removed 75 samples (--mind 0.1) - still not working, now because not enough data. 
```
```{r}
# import eigenvalues
eigenvec <- read_tsv("chloro_pca.eigenvec")
str(eigenvec)
eigenvec$ID <- eigenvec$IID
eigenvec <- eigenvec %>%
  separate(ID, into = c("ID", "species"), sep = "_", remove = T)
                
# plotting the pca
ggplot(eigenvec) +
  geom_point(aes(x = PC1, y = PC2, colour=species)) +
  labs(title = "Chlorplast variants (51 snps)",
       x = "PC1",
       y = "PC2")
```

